src/test/java/hudson/plugins/sshslaves/verifiers/TrileadVersionSupportManagerTest.java

```diff
@@ -1,44 +1,147 @@
-        ClassLoader loader = Thread.currentThread().getContextClassLoader();
-        try {
-            Thread.currentThread().setContextClassLoader(new BlockingClassloader(Thread.currentThread().getContextClassLoader()));
-            String name = TrileadVersionSupportManager.getTrileadSupport().getClass().getName();
-            assertEquals("hudson.plugins.sshslaves.verifiers.TrileadVersionSupportManager$LegacyTrileadVersionSupport", name);
-        } finally {
-            Thread.currentThread().setContextClassLoader(loader);
-        }
+        BlockingClassloader classloader = newBlockingClassloader();
+        classloader.block("com.trilead.ssh2.signature.KeyAlgorithmManager");
+
+        Object trileadSupport = invokeGetTrileadSupport(classloader);
+        assertEquals("hudson.plugins.sshslaves.verifiers.TrileadVersionSupportManager$LegacyTrileadVersionSupport", trileadSupport.getClass().getName());
+    }
+
+    @Test
+    @Issue("JENKINS-44893")
+    public void testLegacyInstanceWithLinkageError() {
+        BlockingClassloader classloader = newBlockingClassloader();
+        classloader.inspectPackage("com.trilead.ssh2.signature");
+        classloader.block("com.trilead.ssh2.signature.KeyAlgorithm");
+
+        Object trileadSupport = invokeGetTrileadSupport(classloader);
+        assertEquals("hudson.plugins.sshslaves.verifiers.TrileadVersionSupportManager$LegacyTrileadVersionSupport", trileadSupport.getClass().getName());
     }
 
     @Test
     public void testCurrentInstance() {
         assertEquals(JenkinsTrilead9VersionSupport.class, TrileadVersionSupportManager.getTrileadSupport().getClass());
     }
 
+    @Test
+    @Issue("JENKINS-44893")
+    public void testCurrentInstanceWithIsolatedClassLoader() {
+        BlockingClassloader classloader = newBlockingClassloader();
+        Object trileadSupport = invokeGetTrileadSupport(classloader);
+
+        assertEquals("hudson.plugins.sshslaves.verifiers.JenkinsTrilead9VersionSupport", trileadSupport.getClass().getName());
+        assertNotEquals(JenkinsTrilead9VersionSupport.class, trileadSupport.getClass());
+    }
+
+    private static Object invokeGetTrileadSupport(ClassLoader classloader) {
+        try {
+            Class<?> clz = Class.forName("hudson.plugins.sshslaves.verifiers.TrileadVersionSupportManager", true, classloader);
+            Method method = clz.getDeclaredMethod("getTrileadSupport");
+            method.setAccessible(true);
+            return method.invoke(null);
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static BlockingClassloader newBlockingClassloader() {
+        BlockingClassloader classloader = new BlockingClassloader(TrileadVersionSupportManagerTest.class.getClassLoader());
+        classloader.inspectPackage("hudson.plugins.sshslaves");
+        return classloader;
+    }
+
+
+    private static class BlockingClassloader extends URLClassLoader {
```

Comment:
Not the standard `MaskingClassLoader`?

Reply:
I did not know there is a 'MaskingClassLoader', I'll look at it 👍

Reply:
@jglick 
I tried MaskingClassLoader but did not get it right. Correct me if I'm wrong.

The MaskingClassLoader uses parent-first strategy when loading class, thus will unable to 'mask' classes referenced by that is loaded with parent classloader. Referrer https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-5.html#jvms-5.3.

The BlockingClassloader uses parent-last strategy. When loading 'inspected' classes/packages, it becomes a 'defining' classloader, and will be able to 'block' any classes as will, including java.lang.Object class, realistic to create an isolated test environment.

In this test case, I'd prefer BlockingClassloader :)

Reply:
The justification LGTM, just a test code anyway

Reply:
ping @jglick